retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/ObservableTest.java

```diff
@@ -103,31 +97,32 @@
   @Test public void responseSuccess200() {
     server.enqueue(new MockResponse().setBody("Hi"));
 
-    BlockingObservable<Response<String>> o = service.response().toBlocking();
-    Response<String> response = o.first();
+    TestSubscriber<Response<String>> subscriber = new TestSubscriber<>();
+    service.response().subscribe(subscriber);
+    Response<String> response = subscriber.getOnNextEvents().get(0);
     assertThat(response.isSuccessful()).isTrue();
     assertThat(response.body()).isEqualTo("Hi");
+    subscriber.assertCompleted();
```

Comment:
SafeSubscriber is an implementation detail, but that is the observable
contract which is not. If we want to test whether the contract is honored
then the tests need ported to use unsafeSubscribe and there's a lot more
cases to cover.

On Tue, Aug 16, 2016 at 1:25 PM Artem Zinnatullin <notifications@github.com>
wrote:

> In
> retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/ObservableTest.java
> <https://github.com/square/retrofit/pull/1977#discussion_r74980378>:
>
> >      assertThat(response.isSuccessful()).isTrue();
> >      assertThat(response.body()).isEqualTo("Hi");
> > +    subscriber.assertCompleted();
>
> Well, it's implementation detail, but ok
> … <#m_-5899499742275413021_>
> On Tue, 16 Aug 2016, 18:36 Jake Wharton, ***@***.***> wrote: In
> retrofit-adapters/rxjava/src/test/java/retrofit2/adapter/rxjava/ObservableTest.java
> <#1977 (comment)
> <https://github.com/square/retrofit/pull/1977#discussion_r74971315>>: >
> assertThat(response.isSuccessful()).isTrue(); >
> assertThat(response.body()).isEqualTo("Hi"); > +
> subscriber.assertCompleted(); I don't see why. SafeSubscriber dictates the
> TestSubscriber will only ever see a completed event or an error event, so
> the lack of a completed event will already fail the test. — You are
> receiving this because you commented. Reply to this email directly, view it
> on GitHub <
> https://github.com/square/retrofit/pull/1977/files/999c72be0468664c711159cd53deff79e62ee4a1#r74971315>,
> or mute the thread <
> https://github.com/notifications/unsubscribe-auth/AA7B3Ka-y4lWlCucjTGE2ck7P99v1m57ks5qgecBgaJpZM4JlZu3>
> .
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/square/retrofit/pull/1977/files/999c72be0468664c711159cd53deff79e62ee4a1#r74980378>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEEEffhge1NT7kNRaaJbGG8QsuarSdAks5qgfKJgaJpZM4JlZu3>
> .
>